OpenRA.Mods.Common/Traits/AutoCarryall.cs

```diff
@@ -102,35 +159,61 @@ void FindCarryableForTransport(Actor self)
 			}
 		}
 
+		protected override void ResolveOrder(Actor self, Order order)
+		{
+			if (order.OrderString != "Stop")
+			{
+				UnderAutoCommand = false;
+				busy = true;
+			}
+
+			base.ResolveOrder(self, order);
+		}
+
 		sealed class FerryUnit : Activity
 		{
 			readonly Actor cargo;
-			readonly Carryable carryable;
-			readonly Carryall carryall;
+			readonly AutoCarryable carryable;
+			readonly AutoCarryall carryall;
 
 			public FerryUnit(Actor self, Actor cargo)
 			{
 				this.cargo = cargo;
-				carryable = cargo.Trait<Carryable>();
-				carryall = self.Trait<Carryall>();
+				carryable = cargo.Trait<AutoCarryable>();
+				carryall = self.Trait<AutoCarryall>();
 			}
 
 			protected override void OnFirstRun(Actor self)
 			{
-				if (!cargo.IsDead && !carryall.IsTraitDisabled)
+				if (carryall.IsTraitDisabled || carryall.Carryable == null || carryall.Carryable.IsDead || !carryall.UnderAutoCommand)
+					Cancel(self, true);
+				else
 					QueueChild(new PickupUnit(self, cargo, 0, carryall.Info.TargetLineColor));
 			}
 
+			public override IEnumerable<TargetLineNode> TargetLineNodes(Actor self)
+			{
+				if (ChildActivity != null)
+					foreach (var n in ChildActivity.TargetLineNodes(self))
+						yield return n;
+			}
+
 			public override bool Tick(Actor self)
 			{
-				// Cargo may have become invalid or PickupUnit cancelled.
-				if (carryall.IsTraitDisabled || carryall.Carryable == null || carryall.Carryable.IsDead)
+				if (IsCanceling)
 					return true;
 
+				// Cargo may have become invalid or PickupUnit cancelled.
+				if (carryall.IsTraitDisabled || carryall.Carryable == null || carryall.Carryable.IsDead || !carryall.UnderAutoCommand)
```

Comment:
```suggestion
				if (carryall.IsTraitDisabled || carryall.Carryable == null || carryall.Carryable.IsDead)
```

please remove this, it does nothing but bug out queued orders specifically it makes this statement incorrect

> 2. if you queue order after FerryUnit pick up the actor, then the carrier will deliver the unit then run the queued

Reply:
No, 2 is the same in test. And remove this line you will make 1 incorrect, it will become the same as 2.
> 1. if you queue order before FerryUnit pick up the actor, then the carrier will pick up unit then run the queued orders

Reply:
and that is the correct behaviour, if the player wants to interrupt the carryall he will simply not hold down shift when giving an order to the carryall

Reply:
In addition, I need to ask in which usecase it 
> it does nothing but bug out queued orders specifically

Reply:
It's a matter of consistency, as for usecase it might be that the player may want to use the carryall for something else after its done carrying the harvester

Reply:
> and that is the correct behaviour, if the player wants to interrupt the carryall he will simply not hold down shift when giving an order to the carryall

It will mess the target line of queue orders if you give up 1. When you are holding shift you see the target line and `FerryUnit` has 2 target lines at different stages:

1. before carryall pick up the harvester, the target line point to the harvester
2. after carryall pick up the harvester, the target line point to the location

In game when holding shift queueing orders, you will see a target line change when carryall pick up the harvester, which will be confusing for player, they may think there is an incorrectly order queued by program. 

In addition, in player's view, auto action should be like manual actions, which divide pick up and deliver into 2 parts, if queue orders after part A, part B can be skipped.

Reply:
I think I find the problem in our discussion: should we keep
> if you queue order before FerryUnit pick up the actor, then the carrier will pick up unit then run the queued orders

I am fine to remove this, it will also simplify the code, but this will have the target line problem I mentioned above.

Reply:
targetline is a separate problem that can be addressed separately

Reply:
As long as the targetline is fixed we can drop 1 now.

Reply:
fixed